### PR TITLE
LibGUI:  Move Action handling from WindowServerConnection to Widget

### PR DESCRIPTION
**Pull request by metmo**
This PR moves the Action execution handling from `WindowServerConnection` to `Widget`.
When a key event gets posted in the event loop it will bubble up through the hierarchy until it reaches the top Widget under Window. Only if the event is marked as ignored when it reaches the last widget it will search for a suitable Action.
This means that it's up to the widget to choose if it wants to accept the event and thereby not trigger any action, or mark the event as ignored to let possible actions be triggered.

This requires some adjustment in the widgets `keydown_event()` since we previously relied on the Action having precedence.

~~This allows a widget to disable actions with certain modifiers while the widget is focused. The main use is TextEditor which previously would miss Key & Shift+Key keypresses when an action used the respective keys as shortcuts.
Exceptions can be added for keys we always need to trigger actions, like F1 or Del.~~

### CODE REVIEW

**Comment by awesomekling** [comment]
Can we instead simply process key events *before* looking for an action?
Events have an `accepted` flag that tells us if some widget used it.
If event processing concludes with the event still unaccepted, we look for an action *then*.
Or maybe I am missing something here?

**Comment by metmo** [comment]
I think I follow along, the reason I put the processing _after_ looking for an action was to reduce the number of lookups and only check for matches when an action _would_ be triggered, since most of the key events won't trigger an action.
Are you suggesting moving the triggering of the action down the event loop?

**Comment by metmo** [comment]
Or are you suggesting sending the key event to `Widget::key_down_event()` for the focused widget before looking for an action?

**Comment by awesomekling** [comment]
> Or are you suggesting sending the key event to `Widget::event()` for the focused widget before looking for an action?

Yes, exactly :)

File: Userland/Applications/PixelPaint/ImageEditor.cpp [hunk 1]
```diff
@@ -260,6 +260,8 @@ void ImageEditor::keydown_event(GUI::KeyEvent& event)
 {
     if (m_active_tool)
         m_active_tool->on_keydown(event);
+
+    event.ignore();
```
Comment (awesomekling) [inline Userland/Applications/PixelPaint/ImageEditor.cpp:264]: I think we should let the active tool decide if the event is accepted.

File: Userland/Games/Solitaire/Game.cpp [hunk 2]
```diff
@@ -231,6 +231,7 @@ void Game::keydown_event(GUI::KeyEvent& event)
     } else if (event.shift() && event.key() == KeyCode::Key_F11) {
         dump_layout();
     }
+    event.ignore();
```
Comment (awesomekling) [inline Userland/Games/Solitaire/Game.cpp:234]: I think this should be in an `else` branch. "Accepted" means that the widget did something meaningful with the event and no further processing is needed.

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 3]
```diff
@@ -827,12 +829,14 @@ void TextEditor::keydown_event(KeyEvent& event)
...
+        event.ignore();
         return;
     }
 
     if (event.key() == KeyCode::Key_Delete) {
         if (m_autocomplete_box)
             m_autocomplete_box->close();
+        event.ignore();
         return;
     }
```
Comment (awesomekling) [inline Userland/Libraries/LibGUI/TextEditor.cpp:841]: These are so weird. Why does text editor not handle `Key_Delete` itself in the first place? 🤔

File: Userland/Libraries/LibVT/TerminalWidget.cpp [hunk 4]
```diff
@@ -243,6 +243,8 @@ void TerminalWidget::keydown_event(GUI::KeyEvent& event)
 
     if (event.key() != Key_Control && event.key() != Key_Alt && event.key() != Key_LeftShift && event.key() != Key_RightShift && event.key() != Key_Super)
         scroll_to_bottom();
+
+    event.ignore();
```
Comment (awesomekling) [inline Userland/Libraries/LibVT/TerminalWidget.cpp:247]: This feels too broad, if we did something with the event, we should accept it, not just pass it up the chain.

File: Userland/Libraries/LibWeb/InProcessWebView.cpp [hunk 5]
```diff
@@ -273,7 +273,7 @@ void InProcessWebView::keydown_event(GUI::KeyEvent& event)
         }
     }
 
-    event.accept();
+    event.ignore();
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/InProcessWebView.cpp:276]: This seems at odds with `page_accepted_event` above. If the page accepted the event, we shouldn't call `ignore()` on it.

File: Userland/Libraries/LibWeb/OutOfProcessWebView.cpp [hunk 6]
```diff
@@ -163,6 +163,7 @@ void OutOfProcessWebView::handle_resize()
 void OutOfProcessWebView::keydown_event(GUI::KeyEvent& event)
 {
     client().async_key_down(event.key(), event.modifiers(), event.code_point());
+    event.ignore();
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/OutOfProcessWebView.cpp:166]: Oh frick this is a really tricky one.. What does this mean for out-of-process browser engines..
We will now send all key events to the web content, even stuff that's potentially UI actions, like Ctrl+T ("open new tab") will become a JavaScript event despite the user *most likely* wanting to open a new tab..

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibGUI/Widget.cpp [hunk 7]
```diff
@@ -308,13 +309,62 @@ void Widget::event(Core::Event& event)
...
+            return true;
+        }
+        if (action->swallow_key_event_when_disabled()) {
+            event.accept();
+            return true;
+        }
+    }
+    return false;
+}
+
```
Comment (metmo) [inline Userland/Libraries/LibGUI/Widget.cpp:357]: Breaking this out to allow Widgets to trigger actions before other events. I'll admit returning a bool here isn't the best looking solution, if anyone has a better idea I'm open for it!

**Review by metmo — COMMENTED** [review]
Third iteration of this one with, hopefully, more clear logic and flow

**Pull request closed without merging**